website/static/js/fileBrowser.js

```diff
@@ -0,0 +1,1175 @@
+                            parentID : self.breadcrumbs()[self.breadcrumbs().length-1].data.id,
+                            modalID : 'addSubcomponent',
+                            stayCallback : function _stayCallback_inPanel() {
+                                self.allProjectsLoaded(false);
+                                self.updateList(self.breadcrumbs()[self.breadcrumbs().length-1]);
+                            }
+                        })
+                    ]));
+                }
+            }
+            // if we have more pages keep loading the pages
+            if (value.links.next) {
+                self.loadingNodePages = true;
+                var collData = {};
+                if(!self.allProjectsLoaded()) {
+                    collData = { systemCollection : 'nodes' };
+                }
+                self.updateList({link : value.links.next, data : collData });
+                return; // stop here so the reloads below don't run
+            } else {
+                self.loadingNodePages = false;
+            }
+            if(self.loadingAllNodes) {
+                self.data().data = _makeTree(self.data().data);
+                self.allProjects(self.data());
+                self.generateFiltersList();
+                self.loadingAllNodes = false;
+                self.allProjectsLoaded(true);
+            }
+            self.reload(true);
+            self.refreshView(false);
+        };
+        self.updateListError = function _updateListError (result){
+            self.nonLoadTemplate(m('.fb-error.text-danger.m-t-lg', [
+                m('p', m('i.fa.fa-exclamation-circle')),
+                m('p','Projects for this selection couldn\'t load.'),
+                m('p', m('.btn.btn-default', {
+                    onclick : self.updateFilter.bind(null, self.systemCollections[0])
+                },' Reload \'All My Projects\''))
+            ]));
+            self.data().data = [];
+            self.refreshView(false);
+            throw new Error('Receiving initial data for File Browser failed. Please check your url');
+        };
+        self.generateFiltersList = function _generateFilterList () {
+            self.users = {};
+            self.tags = {};
+            self.data().data.map(function _generateFiltersListMap(item){
+                var contributors = item.embeds.contributors.data ? item.embeds.contributors.data : [];
+                for(var i = 0; i < contributors.length; i++) {
+                    var u = contributors[i];
+                    if(self.users[u.id] === undefined) {
+                        self.users[u.id] = {
+                            data : u,
+                            count: 1
+                        };
+                    } else {
+                        self.users[u.id].count++;
+                    }
+                }
+
+                var tags = item.attributes.tags ? item.attributes.tags : [];
+                for(var j = 0; j < tags.length; j++) {
```

Comment:
There's duplication here with L344-354 which can be refactored, but it's not absolutely necessary.